### PR TITLE
Fix surrogates parsing on regex

### PR DESCRIPTION
**Pull request by jedel1043**
Just a simple fix to ensure unpaired surrogates are correctly handled instead of ignored.
Also modified the range check for surrogates, since we just want to check if the first surrogate is high; if it is low, it is already unpaired, so we can directly return.

### CODE REVIEW

File: src/parse.rs [hunk 1]
```diff
@@ -870,34 +870,29 @@ where
             }
             match u16::from_str_radix(&s, 16) {
                 Ok(u) => {
-                    if (0xDC00..=0xDFFF).contains(&u) || (0xD800..=0xDB7F).contains(&u) {
-                        // Low/High Surrogates
+                    if (0xD800..=0xDBFF).contains(&u) {
+                        // Low Surrogates
```
Comment (ridiculousfish) [inline src/parse.rs:874]: These are actually the high surrogates 🤷

File: src/parse.rs [hunk 2]
```diff
@@ -870,34 +870,29 @@ where
...
                         if !self.try_consume_str("\\u") {
                             return Some(u as u32);
                         }
+                        orig_input = self.input.clone();
 
-                        let mut s = String::new();
-                        for _ in 0..4 {
-                            if let Some(c) = self.next().and_then(char::from_u32) {
+                        let result = (|| {
```
Comment (ridiculousfish) [inline src/parse.rs:880]: I haven't seen this "immediately invoked function expression" pattern used before in Rust. Any reason not to just use a `{...}` block?

**Review by ridiculousfish — CHANGES_REQUESTED** [review]
Thank you for this. I think this is more faithfully implementing the behavior of `RegExpUnicodeEscapeSequence` in [the grammar](https://262.ecma-international.org/6.0/#sec-patterns) so I think it's right, but since you've thought about it more recently, can you please share an example regexp which did the wrong thing before?

File: src/parse.rs (hunk 1 above)
Comment (jedel1043) [reply src/parse.rs:874]: Ah, my comment was to signal that we're trying to parse the low surrogates next 😅
I'll add a more descriptive comment

File: src/parse.rs (hunk 2 above)
Comment (jedel1043) [reply src/parse.rs:880]: Ah, this is just a poor's man try block 😆 it allows to use `?` inside a block without returning from the main function.

I had another version that used combinators instead, but it was a bit less readable than this imperative version.

**Comment by jedel1043** [comment]
> can you please share an example regexp which did the wrong thing before?

Both new tests are examples of regular expressions that failed before the change, but some simple examples would be any succession of two unpaired surrogates:

```rust
Regex::new(r"(?:[\uD800\uDBFF])").unwrap_err();
Regex::new(r"(?:[\uDC00\uDFFF])").unwrap_err();
```

**Comment by ridiculousfish** [comment]
Makes sense, thank you! Merged.

**Pull request merged**